File: calculator/calculator.py
def add(summand_a, summand_b):
    return summand_a + summand_b


def subtract(minuend, subtrahend):
    return minuend - subtrahend


def factorial(x):
    if x < 0:
        raise ValueError("factorial undefined for negative values.")
    result = 1
    for i in range(1, x+1):
        result *= i
    return result

File: src/main/java/tech/bison/trainee/Calculator.java
package tech.bison.trainee;

import static java.util.stream.IntStream.rangeClosed;

public class Calculator {
  public int add(int summandA, int summandB) {
    return summandA + summandB;
  }

  public int subtract(int summandA, int summandB) {
    return summandA - summandB;
  }

  public int factorial(int x) {
    if (x < 0) {
      throw new UnsupportedOperationException("factorial undefined for negative values.");
    } else {
      return rangeClosed(1, x).reduce(1, (a, b) -> a * b);
    }
  }

  public int multiply(int a, int b) {
    return a * b;
  }

  public double divide(double a, double b) {
    if (b > 0) {
      return a / b;
    } else {
      throw new UnsupportedOperationException("division undefined for divisor zero.");
    }
  }
}
